### PR TITLE
Update Rivet (and YODA) builds with an eye toward Rivet integration into O2 pipeline 

### PR DESCRIPTION
**Pull request by cholmcc**
This will do the following updates 
- YODA and Rivet taken directly from upstream repositories.   The older `alisw` repositories are terribly out of date, and the local fixes are no longer needed. 
- YODA depends on HepMC3 to make sure we pull in the right ROOT version (HepMC3 depends on ROOT) when making Rivet 
- Rivet built against HepMC3 as needed for Rivet integration into O2 (see also https://gitlab.com/cholmcc/O2Rivet)

The changes have been checked and validated on `lxplus` albeit in a slightly funny way (because `aliBuild` doesn't play nice with CVMFS).  However, the tests are entirely equivalent to how `aliBuild` would do it from scratch. 

Yours,
_Christian_

### CODE REVIEW

**Comment by ktf** [comment]
Looks good to me. Pinging @alisw/pagmc for actual approval.

**Comment by cholmcc** [comment]
Hi, just a fix - the build failed because of a typo.  Hopefully last commit will fix.

**Comment by jackal1-66** [comment]
@cholmcc did you test if AliGenerators builds and runs correctly on lxplus with your additions? The edits look good to me, but it's a matter of understanding also if the system will work both on AliGenerators and on the GRID without using O2.

File: rivet.sh [hunk 1]
```diff
@@ -1,76 +1,147 @@
 package: Rivet
 version: "%(tag_basename)s"
-tag: "3.1.6-alice1"
-source: https://github.com/alisw/rivet
+tag: "rivet-3.1.8"
+source: https://gitlab.com/hepcedar/rivet.git
```
Comment (jackal1-66) [inline rivet.sh:4]: Some modifications were done in the past to make RIVET work on the ALICE framework, I fear this force pull from the hepcedar repo will break some stuff. @maireiphc is definitely more informed than me on this

Comment (jackal1-66) [inline rivet.sh:4]: Some modifications were done in the past to make RIVET work on the ALICE framework, I fear this force pull from the hepcedar repo will break some stuff. @maireiphc is definitely more informed than me on this

File: rivet.sh [hunk 2]
```diff
@@ -1,76 +1,147 @@
...
   - "Python:(?!osx)"
   - "Python-modules:(?!osx)"
   - "Python-system:(osx.*)"
 build_requires:
   - GCC-Toolchain:(?!osx)
+  - YODA
+  - Python
 prepend_path:
-  PYTHONPATH: $RIVET_ROOT/lib/python3.9/site-packages
+  PYTHONPATH: $RIVET_ROOT/lib/python/site-packages
```
Comment (jackal1-66) [inline rivet.sh:18]: What about systems that contain multiple python versions and have a python2 default? this could break building...

File: yoda.sh [hunk 3]
```diff
@@ -1,107 +1,86 @@
 package: YODA
 version: "%(tag_basename)s"
-tag: "v1.9.5"
-source: https://github.com/alisw/yoda
+tag: "yoda-1.9.7"
+source: https://gitlab.com/hepcedar/yoda.git
```
Comment (jackal1-66) [inline yoda.sh:4]: Same comment as for rivet.sh

**Review by jackal1-66 — COMMENTED** [review]
I reviewed the part I have more experience in, I leave the rest (mostly the yoda part) to @maireiphc

**Comment by cholmcc** [comment]
> @cholmcc did you test if AliGenerators builds and runs correctly on lxplus with your additions? 

Most EGs do no require YODA nor Rivet (which are the only ones changed).   

Here's the situation 

  - AMPT - does not use Rivet 
  - CRMC - Works with newer Rivet - I know, I wrote the code for CRMC :-) 
  - DPMJET - does not use Rivet 
  - EPOS - does not use Rivet 
  - Herwig:(?!osx) - does not use Rivet 
  - JEWEL - does not use Rivet 
  - POWHEG - does not use Rivet
  - pythia - does not use Rivet  directly (it can, but not configured for it)
  - pythia6 - does not use Rivet
  - SHERPA - does not use Rivet. It can, but not configured for it.  If it was, it would work with newer Rivet too
  - ThePEG - uses Rivet and will most likely work with newer Rivet too. 
  - AGILe:(?!osx) - should work with newer Rivet - though not configured for it
  - Sacrifice -  should work with newer Rivet - though not configured for it
  - aligenmc - does not use Rivet
  - FONLL - does not use Rivet
  - Therminator2 - does not use Rivet
  - Rivet - Well :-)
  - lhapdf-pdfsets - does not use Rivet
  - JETSCAPE - does not use Rivet
  - EPOS4 - does not use Rivet

It is only ThePEG that could have problems, but it looks unlikely.   

Argh, I made a mistake - another stupid typo (`CGAL_LDFAGS` should have been `CGAL_LDF`_L_`AGS`).  - Retry.

**Comment by cholmcc** [comment]
From `AliGenerators` build log: 

```
## sw/BUILD/ThePEG-latest/log
--
  108 |       _axis = p2._axis;
      |                  ^~~~~
NLORivetAnalysis.cc: In member function 'virtual void ThePEG::NLORivetAnalysis::analyze(ThePEG::tEventPtr, long int, int, int)':
NLORivetAnalysis.cc:128:22: error: no matching function for call to 'Rivet::AnalysisHandler::analyze(HepMC::GenEvent&)'
  128 |       _rivet->analyze(*hepmc);
      |       ~~~~~~~~~~~~~~~^~~~~~~~
/sw/slc7_x86-64/Rivet/rivet-3.1.8-local1/include/Rivet/AnalysisHandler.hh:239:10: note: candidate: 'void Rivet::AnalysisHandler::analyze(const HepMC3::GenEvent&)'
```

This is a problem I've seen reported elsewhere, that somehow the compiler fails to promote a `HepMC3::GenEvent&` to `const HepMC3::GenEvent&`.   It seems to be a compiler issue - not a code issue.    The fix seems to be to explicitly cast the object to be `const` 

```
_rivet->analyze(std::const_cast<const HepMC3::GenEvent&>(*hepmc));
```

four places in `Analysis/NLORivetAnalysis.cc` and one place in `Analysis/RivetAnalysis.cc`.   in the ThePEG source directory. 

Note, with newer GCC (I use 13.2) it is not an issue. 

Yours,
_Christian_

**Comment by cholmcc** [comment]
What ever errors came from `build/O2/alidist-cs8-perftest` are _not_ from the changes in this MR.  Looking at the full log it seems that what failed was the test 

```
The following tests FAILED:
	  1 - o2-qc-test-core (Timeout)
``` 

so a false negative.

**Comment by cholmcc** [comment]
The test `build/alidist/O2Suite/o2/macOS` failed because of a problem with `curl` - not this MR - a false negative.

**Comment by cholmcc** [comment]
`build/alidist/O2Suite/o2/macOS-arm ` actually did succeed - another false negative

**Comment by cholmcc** [comment]
> From `AliGenerators` build log:
...
> I've made two PR's against AliGenerator to fix this issue - [here](https://github.com/alisw/thepeg/pull/4) and [here](https://github.com/alisw/thepeg/pull/5)
> 

Any news on this?  Thank you. 

Yours, 
_Christian_

**Comment by cholmcc** [comment]
> @cholmcc the issue is, however for the packages that require Rivet, like AliGenerators, which is still largely used. Losing the ability to run on the GRID with next releases is not what we're looking for at the moment. Antonin is more informed on the update to Rivet 3.1.6 on our repository. I will ping him to answer here. @maireiphc

Note the two other merge requests against ThePEG ([here](https://github.com/alisw/thepeg/pull/4) and [here](https://github.com/alisw/thepeg/pull/5)) - they should fix all issues of this MR wrt to `AliGenerators`.  Thus, if you accept those two MRs and then there should be no problem for the next release of either `AliGenerators` or `alidist`. 

WRT usage of Rivet - especially from `AliGenerators` packages.  It seems unlikely that any user uses Rivet directly from something like ThePEG, and in some sense that shouldn't be the use of Rivet for ALICE user.  Rather, ALICE users should either, 
- use the upcoming [O2Rivet](https://gitliab.com/cholmcc/O2Rivet) - see also meeting [Thursday](https://indico.cern.ch/event/1334550/) - on existing simulated data, or 
- generate HepMC data files with some EG (typically locally) and run Rivet directly on that. 

Therefore, it makes less sense to build an `AliGenerator` with Rivet support _directly_ linked in.  Most EGs that support such an option (CRMC, Pythia8, Herwig, ...) does so for the benefit of developers and less so for the users.  Thus, Rivet shouldn't be switch on as a compile-time option for EGs unless explicitly needed by the EG. 

In summary, please consider to apply the MRs to ThePEG and this MR to `alidist`, then all should be good. 

Yours,
_Christian_

**Comment by jackal1-66** [comment]
@cholmcc the related PRs were merged. I have no more comments on this. But the system will be tested on the GRID as soon as the new AliGen version is released.

**Comment by jackal1-66** [comment]
@ktf can we restart the checks somehow after the merging of the PRs linked by Christian?

**Review by jackal1-66 — APPROVED** [review]
Everything looks good. Approved for merging.

**Comment by ktf** [comment]
This is still broken in the generators test. I guess you need proper tags for the changes you proposed in a couple of generators?

**Comment by jackal1-66** [comment]
Tagging @cholmcc to check this

**Comment by cholmcc** [comment]
Hi all, 

Sorry for the radio silence - off for the "potato" holiday the week before last and busy with other stuff last week. 

I see that `build/AliGenerators/generators` test fail with failure in the `thePEG`:  

 ```
--
  127 |     struct _GLIBCXX17_DEPRECATED iterator
      |                                  ^~~~~~~~
NLORivetAnalysis.cc: In member function 'virtual void ThePEG::NLORivetAnalysis::analyze(ThePEG::tEventPtr, long int, int, int)':
NLORivetAnalysis.cc:128:22: error: no matching function for call to 'Rivet::AnalysisHandler::analyze(HepMC::GenEvent&)'
  128 |       _rivet->analyze(*hepmc);
      |       ~~~~~~~~~~~~~~~^~~~~~~~
/sw/slc7_x86-64/Rivet/rivet-3.1.8-local4/include/Rivet/AnalysisHandler.hh:239:10: note: candidate: 'void Rivet::AnalysisHandler::analyze(const HepMC3::GenEvent&)'
--
/sw/slc7_x86-64/Rivet/rivet-3.1.8-local4/include/Rivet/AnalysisHandler.hh:245:34: note:   no known conversion for argument 1 from 'HepMC::GenEvent' to 'const HepMC3::GenEvent*'
  245 |     void analyze(const GenEvent* event);
      |                  ~~~~~~~~~~~~~~~~^~~~~
NLORivetAnalysis.cc:152:26: error: no matching function for call to 'Rivet::AnalysisHandler::analyze(HepMC::GenEvent&)'
  152 |           _rivet->analyze(*hepmc);
      |           ~~~~~~~~~~~~~~~^~~~~~~~
/sw/slc7_x86-64/Rivet/rivet-3.1.8-local4/include/Rivet/AnalysisHandler.hh:239:10: note: candidate: 'void Rivet::AnalysisHandler::analyze(const HepMC3::GenEvent&)'
--
  245 |     void analyze(const GenEvent* event);
``` 

This was addressed by the two MRs [here](https://github.com/alisw/thepeg/pull/4) and [here](https://github.com/alisw/thepeg/pull/5), which _have_ been merged.   Thus, it seems like the test `build/AliGenerators/generators`  pulls a version of [thePEG](https://github.com/alisw/thepeg) which does _not_ contain the two fixes.  

- [`aligenerators.sh`](https://github.com/alisw/alidist/blob/master/aligenerators.sh) simply depends on **ThePEG**.
- [`thepeg.sh`](https://github.com/alisw/alidist/blob/master/thepeg.sh) pulls the tag [`v2.2.2-alice1`](https://github.com/alisw/thepeg/tree/v2.2.2-alice1) which _does not_ contain the two above mentioned fixes. 

The **solution** as I see it: 

1. Create new tag on **ThePEG** from current [`master`](https://github.com/alisw/thepeg) - say `v2.2.2-alice2`
2. Update [`thepeg.sh`](https://github.com/alisw/alidist/blob/master/thepeg.sh)  to pull that tag 
3. Re-test 

Yours,
_Christian_

**Comment by jackal1-66** [comment]
> Hi all,
> 
> Sorry for the radio silence - off for the "potato" holiday the week before last and busy with other stuff last week.
> 
> I see that `build/AliGenerators/generators` test fail with failure in the `thePEG`:
> 
> ```
> --
>  127 |     struct _GLIBCXX17_DEPRECATED iterator
>      |                                  ^~~~~~~~
> NLORivetAnalysis.cc: In member function 'virtual void ThePEG::NLORivetAnalysis::analyze(ThePEG::tEventPtr, long int, int, int)':
> NLORivetAnalysis.cc:128:22: error: no matching function for call to 'Rivet::AnalysisHandler::analyze(HepMC::GenEvent&)'
>  128 |       _rivet->analyze(*hepmc);
>      |       ~~~~~~~~~~~~~~~^~~~~~~~
> /sw/slc7_x86-64/Rivet/rivet-3.1.8-local4/include/Rivet/AnalysisHandler.hh:239:10: note: candidate: 'void Rivet::AnalysisHandler::analyze(const HepMC3::GenEvent&)'
> --
> /sw/slc7_x86-64/Rivet/rivet-3.1.8-local4/include/Rivet/AnalysisHandler.hh:245:34: note:   no known conversion for argument 1 from 'HepMC::GenEvent' to 'const HepMC3::GenEvent*'
>  245 |     void analyze(const GenEvent* event);
>      |                  ~~~~~~~~~~~~~~~~^~~~~
> NLORivetAnalysis.cc:152:26: error: no matching function for call to 'Rivet::AnalysisHandler::analyze(HepMC::GenEvent&)'
>  152 |           _rivet->analyze(*hepmc);
>      |           ~~~~~~~~~~~~~~~^~~~~~~~
> /sw/slc7_x86-64/Rivet/rivet-3.1.8-local4/include/Rivet/AnalysisHandler.hh:239:10: note: candidate: 'void Rivet::AnalysisHandler::analyze(const HepMC3::GenEvent&)'
> --
>  245 |     void analyze(const GenEvent* event);
> ```
> 
> This was addressed by the two MRs [here](https://github.com/alisw/thepeg/pull/4) and [here](https://github.com/alisw/thepeg/pull/5), which _have_ been merged. Thus, it seems like the test `build/AliGenerators/generators` pulls a version of [thePEG](https://github.com/alisw/thepeg) which does _not_ contain the two fixes.
> 
> * [`aligenerators.sh`](https://github.com/alisw/alidist/blob/master/aligenerators.sh) simply depends on **ThePEG**.
> * [`thepeg.sh`](https://github.com/alisw/alidist/blob/master/thepeg.sh) pulls the tag [`v2.2.2-alice1`](https://github.com/alisw/thepeg/tree/v2.2.2-alice1) which _does not_ contain the two above mentioned fixes.
> 
> The **solution** as I see it:
> 
> 1. Create new tag on **ThePEG** from current [`master`](https://github.com/alisw/thepeg) - say `v2.2.2-alice2`
> 2. Update [`thepeg.sh`](https://github.com/alisw/alidist/blob/master/thepeg.sh)  to pull that tag
> 3. Re-test
> 
> Yours, _Christian_

Created new tag of ThePEG and PR opened for the recipe update on alidist.

**Comment by cholmcc** [comment]
Hmm, so it seems like **ThePEG** is picking up _both_ HepMC _and_ HepMC3 but cannot decide between which one to use.   

```
NLORivetAnalysis.cc: In member function 'virtual void ThePEG::NLORivetAnalysis::analyze(ThePEG::tEventPtr, long int, int, int)':
NLORivetAnalysis.cc:128:22: error: no matching function for call to 'Rivet::AnalysisHandler::analyze(const HepMC::GenEvent&)'
  128 |       _rivet->analyze(const_cast<const HepMC::GenEvent&>(*hepmc));
      |       ~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/sw/slc7_x86-64/Rivet/rivet-3.1.8-local1/include/Rivet/AnalysisHandler.hh:239:10: note: candidate: 'void Rivet::AnalysisHandler::analyze(const HepMC3::GenEvent&)'
--
/sw/slc7_x86-64/Rivet/rivet-3.1.8-local1/include/Rivet/AnalysisHandler.hh:245:34: note:   no known conversion for argument 1 from 'const HepMC::GenEvent' to 'const HepMC3::GenEvent*'
  245 |     void analyze(const GenEvent* event);
      |                  ~~~~~~~~~~~~~~~~^~~~~
NLORivetAnalysis.cc:152:26: error: no matching function for call to 'Rivet::AnalysisHandler::analyze(const HepMC::GenEvent&)'
  152 |           _rivet->analyze(const_cast<const HepMC::GenEvent&>(*hepmc));
      |           ~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/sw/slc7_x86-64/Rivet/rivet-3.1.8-local1/include/Rivet/AnalysisHandler.hh:239:10: note: candidate: 'void Rivet::AnalysisHandler::analyze(const HepMC3::GenEvent&)'
```

Note in particular 

```
 no known conversion for argument 1 from 'const HepMC::GenEvent' to 'const HepMC3::GenEvent*'
```

Looking more carefully over the [log](https://ali-ci.cern.ch/alice-build-logs/alisw/alidist/5171/a572dc59a593b7ce542384496e2a88600c84326d/build_AliGenerators_generators/fullLog.txt) of the build, I see that the **ThePEG** `configure` script picks up `HepMC` (i.e., version 2)

```
*** HepMC:		/sw/slc7_x86-64/HepMC/HEPMC_02_06_10-39
``` 

and does _not_ see HepMC3 in the environment.   However, since **Rivet** _requires__ HepMC3 (in `/sw/slc7_x86-64/HepMC3/3.2.5-99/`) it is added to the environment, and so there is a conflict (**ThePEG** _thinks_ its building against HepMC while **Rivet** forces HepMC3.).  

The fix seems to me to update [`thepeg.sh`](https://github.com/alisw/alidist/blob/master/thepeg.sh) to require `HepMC3` rather than `HepMC` - or perhaps take out the explicit dependency and let the `Rivet` dependency pull in `HepMC3` - if that works.  I made an [MR](https://github.com/alisw/alidist/pull/5206).

Yours,
_Christian_

File: rivet.sh (hunk 2 above)
Comment (maireiphc) [reply rivet.sh:18]: In the recipe alidist/yoda.sh, there is already a gymnastic arounfd the python site-packages that is in place.

>  # See below at build time the management towards generic path .../lib/python/site-packages
>  PYTHONPATH: $YODA_ROOT/lib/python/site-packages

i.e. 
- comments https://github.com/alisw/alidist/blame/master/yoda.sh#L33 to l.70, making the need of the line " unset PYTHON_VERSION" in yoda.sh l.32
- https://github.com/alisw/alidist/blame/master/yoda.sh#L88 to l.104 for the creation of a "generic" symlink to "python3.x" to "Python"

It was done for yoda.sh in https://github.com/alisw/alidist/commit/fa91a6b879334decaff1dfee55de6058a63820c9

File: rivet.sh
Comment (jackal1-66) [inline rivet.sh:1]: @maireiphc Just not to continue to be stuck on this, I merged the PR and we're going to check today how AliGenerators will behave. If we spot issues we revert it.

Comment (maireiphc) [reply rivet.sh:1]: Ok, let's give a shot indeed.
To be kept in mind : the "right" python gymnastic with symlink is in place for yoda.sh since a ~year but maybe not for rivet.sh. To be seen.

Comment (jackal1-66) [reply rivet.sh:1]: @maireiphc @cholmcc AliGenerators compiled without issues it seems, however when loading it on lxplus I get `this message ROOT/(57):ERROR:105: Unable to locate a modulefile for 'ROOT/'`. I get this also when only loading Rivet by itself. Running on the GRID returns ERROR_IB, as if something is missing, which makes sense considering the issue. So at the moment the new version of Rivet is broken, could you please take care of that @cholmcc? If not I will revert the commit. I tested yesterday's version without the merge and everything is running fine
